tensorly/backend/core.py

```diff
@@ -0,0 +1,854 @@
+import inspect
+import importlib
+import os
+import sys
+import threading
+import types
+import warnings
+
+import numpy as np
+import scipy.linalg
+import scipy.sparse.linalg
+
+
+_STATE = threading.local()
+_DEFAULT_BACKEND = None  # changed in `_initialize_default_backend`
+_KNOWN_BACKENDS = ('numpy', 'mxnet', 'pytorch', 'tensorflow', 'cupy')
+_LOADED_BACKENDS = {}
+
+
+def _initialize_default_backend():
+    """Setup the initial Tensorly backend"""
+    default = 'numpy'
+    backend = os.environ.get('TENSORLY_BACKEND', default)
+
+    if backend not in _KNOWN_BACKENDS:
+        msg = ("TENSORLY_BACKEND should be one of {%s}, got %r. Defaulting to "
+               "'%r'") % (', '.join(map(repr, _KNOWN_BACKENDS)), backend, default)
+        warnings.warn(msg, UserWarning)
+        backend = default
+
+    set_backend(backend, make_default=True)
+
+
+class set_backend(object):
+    """Sets the backend for TensorLy.
+
+    Can be used to set the backend globally, or locally as a contextmanager.
+
+    Parameters
+    ----------
+    backend : {'numpy', 'mxnet', 'pytorch', 'tensorflow', 'cupy'}
+        The name of the backend to use. Default is 'numpy'.
+    make_default : bool, optional
+        If True, the backend will become the default backend for all threads.
+        Note that this only affects threads where the backend hasn't already
+        been explicitly set. If False (default) the backend is only set
+        contextually for the local thread.
+
+    Examples
+    --------
+    Set the backend to numpy globally for this thread:
+
+    >>> import tensorly as tl
+    >>> tl.set_backend('numpy')
+
+    Set the backend to numpy globally for all threads:
+
+    >>> tl.set_backend('numpy', make_default=True)
+
+    Set the backend to use pytorch inside a context:
+
+    >>> with tl.set_backend('pytorch'):  # doctest: +SKIP
+    ...     pass
+    """
+    def __init__(self, backend, make_default=False):
```

Comment:
Not sure I see the point of `make_default` since anyway it will be local to that session -- while I'd expect persistence as a user?
Not sure in which cases we'd care about thread locality.

Reply:
Calling `tl.set_backend` only sets the backend globally for the calling thread (it's a thread local). New threads inherit the default backend unless/until `tl.set_backend` is called explicitly in that thread. For users using threading (in e.g. dask) you may want to set the backend for *all* threads in one call - in that case you can use `make_default` to change the default backend for all threads:

```python
tl.set_backend('pytorch')

# All operation in *this* thread now uses pytorch

# Newly created threads will still default to tensorly's default
# backend, as `set_backend` hasn't been called yet in that thread.

# To get around this, you can change the (not thread local) default
# backend by setting `make_default=True`.

tl.set_backend('pytorch', make_default=True)

# Now all threads that haven't explicitly set a backend themselves will use `pytorch`
```

I think this distinction (and the default value of `False`) makes sense, but it could possibly use a better keyword name than `make_default`. Alternatively, this could be better as a new function `tl.set_default_backend('pytorch')`, which can't be used as a context manager (as that doesn't make sense in this case). Thoughts?

Reply:
what about `make_session_default`? I personally would simply have the backend be global to the session..